SCClassLibrary/Common/Streams/EventStreamCleanup.sc

```diff
@@ -1,47 +1,86 @@
+// The "new" cleanup framework: built-in mechanism to avoid duplicate
+// execution of cleanups, difficult to avoid otherwise due to their
+// replicated nature in a stream graph. Instead of storing bare Functions,
+// they are now stored in CleanupThunks.
+
+// Unlike Thunks which take no arguments, CleanupThunks do allow arguments
+// to be passed to the function, but the CleanupThunk contract is that any
+// subsequent calls into the function would monotonically do the same or
+// less (cleanup) work, so that the first call does the maximal amount of
+// relevant work, making subsequent calls completely redundant.
+// Prototypical example: if a cleanup is called with a "flag: false",
+// it is assumed it will *not* be called later again with "flag: true",
+// if the latter means that the function would have to do "extra" work
+// that was skipped/excluded by "flag: false".
+// CleanupThunk is not a general-purpose function memoization class for
+// functions with arguments.
+CleanupThunk : Thunk {
+
+	value { |... args| ^this.prEvaluate(\valueArray, args) }
+	valueArray { |... args| ^this.prEvaluate(\valueArray, *args) }
+	valueEnvir { |... args| ^this.prEvaluate(\valueArrayEnvir, args) }
+	valueArrayEnvir { |... args| ^this.prEvaluate(\valueArrayEnvir, *args) }
+
+	prEvaluate { |selector, args|
```

Comment:
Instead of adding a new method `prEvaluate`, why not just override `value` from the superclass? Then you can skip all the implementations above (`valueArray`,  `valueEnvir`, `valueArrayEnvir`)

Reply:
> Then you can skip all the implementations above (valueArray, valueEnvir, valueArrayEnvir)

I think this would be incorrect...? Unless I'm missing something.

A CallOnce should be valid to use basically everywhere that a function can be used (with the possible exception of Routine -- I think coroutines are fundamentally incompatible with  Thunk-like behavior). If someone can call `myFunction.valueEnvir` and get parameters from the current environment, then they should also be able to call `CallOnce(myFunction).valueEnvir` and get parameters from the current environment. Flattening all of them into one `value` method would break that.

If you've got another way to do it, perhaps a code example would explain your meaning better.